front/components/agent_builder/capabilities/knowledge/KnowledgeConfigurationSheet.tsx

```diff
@@ -186,54 +193,59 @@ function KnowledgeConfigurationSheetContent({
       description: config.configPageDescription,
       icon: config.icon,
       content: (
-        <FormProvider form={form} onSubmit={handleSave}>
-          <div className="space-y-6">
-            {config.hasTimeFrame && (
-              <TimeFrameSection
-                actionType={
-                  config.name === "extract_data" ? "extract" : "include"
-                }
-              />
-            )}
-            {config.hasJsonSchema && (
-              <JsonSchemaSection
-                initialSchemaString={
-                  action && getJsonSchema(action)
-                    ? JSON.stringify(getJsonSchema(action), null, 2)
-                    : null
-                }
-                agentInstructions={instructions}
-                owner={owner}
-              />
-            )}
-            <DescriptionSection {...config.descriptionConfig} />
-          </div>
-        </FormProvider>
+        <div className="space-y-6">
+          {config.hasTimeFrame && (
+            <TimeFrameSection
+              control={control}
+              actionType={
+                config.name === "extract_data" ? "extract" : "include"
+              }
+            />
+          )}
+          {config.hasJsonSchema && (
+            <JsonSchemaSection
+              control={control}
+              initialSchemaString={
+                action && getJsonSchema(action)
+                  ? JSON.stringify(getJsonSchema(action), null, 2)
+                  : null
+              }
+              agentInstructions={instructions}
+              owner={owner}
+            />
+          )}
+          <DescriptionSection control={control} {...config.descriptionConfig} />
+        </div>
       ),
     },
   ];
 
+  const { isDirty } = useFormState({ control });
+
+  const sources = useWatch({
+    control,
+    name: "sources",
+  });
+  const disableNext = useMemo(() => {
+    if (currentPageId === CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION) {
+      return sources.in.length <= 0 && sources.notIn.length <= 0;
+    }
+    return false;
+  }, [currentPageId, sources.in.length, sources.notIn.length]);
+
   return (
     <MultiPageSheetContent
       pages={pages}
       currentPageId={currentPageId}
       onPageChange={handlePageChange}
       onSave={async (e) => {
         e.preventDefault();
-
-        const isValid = await form.trigger();
-
-        if (isValid) {
-          await form.handleSubmit(handleSave)();
-        }
+        onSave();
       }}
       size="lg"
       showNavigation
-      disableNext={
-        currentPageId === CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION &&
-        !hasDataSources
-      }
-      disableSave={!hasDataSources}
+      disableNext={disableNext}
+      disableSave={!isDirty}
```

Comment:
note: I used `isDirty` instead of `isValid`. Because when on the second page, nothing visually tels the user that only the description is mandatory. So  it doesn't help them having the "save" button disabled. 

By using `isDirty`, when creating a new action, they first chose sources, which makes it dirty, then gets to the timer and description part, button is available, they click on it and now get the error saying they should fill out a description.

Works as well for update. If you open the modal and don't change anything, make sense to having the btn disabled. And if you update something, good chance of the previous values were valid, so ok to let you validate the form.